packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,493 @@
+   * @returns True if the account is a non-EVM account, false otherwise.
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) &&
+      // Non-EVM accounts are backed by a Snap for now
+      account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Handles changes when a new account has been added.
+   *
+   * @param account - The new account being added.
+   */
+  async #handleOnAccountAdded(account: InternalAccount): Promise<void> {
+    if (!this.#isNonEvmAccount(account)) {
+      // Nothing to do here for EVM accounts
+      return;
+    }
+
+    // Get assets list
+    if (account.metadata.snap) {
+      const assets = await this.#getAssetsList(
+        account.id,
+        account.metadata.snap.id,
+      );
+      await this.#refreshAssetsMetadata(assets);
+      this.update((state) => {
+        state.allNonEvmTokens[account.id] = assets;
+      });
+    }
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The new account id being removed.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    // Check if accountId is in allNonEvmTokens and if it is, remove it
+    if (this.state.allNonEvmTokens[accountId]) {
+      this.update((state) => {
+        delete state.allNonEvmTokens[accountId];
+      });
+    }
+  }
+
+  /**
+   * Refreshes the assets snaps and metadata for the given list of assets
+   *
+   * @param assets - The assets to refresh
+   */
+  async #refreshAssetsMetadata(assets: CaipAssetType[]) {
+    const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+      (asset) => !this.state.metadata[asset],
+    );
+
+    // call the snap to get the metadata
+    if (assetsWithoutMetadata.length > 0) {
+      // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+      if (
+        !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+          const chainId = parseCaipAssetType(asset);
+          return Boolean(this.#getAssetSnapFor(chainId));
+        })
+      ) {
+        this.#snaps = this.#getAssetSnaps();
+      }
+      await this.#updateAssetsMetadata(assetsWithoutMetadata);
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const chainId = parseCaipAssetType(asset);
+      if (!assetsByScope[chainId]) {
+        assetsByScope[chainId] = [];
+      }
+      assetsByScope[chainId].push(asset);
+    }
+
+    let newMetadata: Record<CaipAssetType, AssetMetadata>;
+    for (const chainId in assetsByScope) {
+      if (Object.prototype.hasOwnProperty.call(assetsByScope, chainId)) {
```

Comment:
```suggestion
      if (hasProperty(assetsByScope, chainId)) {
```
This exists in `utils`

Reply:
Thanks! here [6e738c9](https://github.com/MetaMask/core/pull/5138/commits/6e738c94845f2a901914a55cc6443f80df3d3c07)